d2renderers/d2svg/d2svg.go

```diff
@@ -746,6 +746,128 @@ func render3dRect(targetShape d2target.Shape) string {
+	border.Stroke = borderStroke
+	borderStyle := targetShape.CSSStyle()
+	border.Style = borderStyle
+	renderedBorder := border.Render()
+
+	var mainPoints []string
+	for _, v := range []d2target.Point{
+		{X: scale(targetShape.Width, 0.25), Y: 0},
+		{X: scale(targetShape.Width, 0.75), Y: 0},
+		{X: targetShape.Width, Y: scale(targetShape.Height, halfYFactor)},
+		{X: scale(targetShape.Width, 0.75), Y: targetShape.Height},
+		{X: scale(targetShape.Width, 0.25), Y: targetShape.Height},
+		{X: 0, Y: scale(targetShape.Height, halfYFactor)},
+	} {
+		mainPoints = append(mainPoints,
+			fmt.Sprintf("%d,%d", v.X+targetShape.Pos.X, v.Y+targetShape.Pos.Y),
+		)
+	}
+
+	mainPointsPoly := strings.Join(mainPoints, " ")
+	// create mask from border stroke, to cut away from the shape fills
+	maskID := fmt.Sprintf("border-mask-%v", svg.EscapeText(targetShape.ID))
+	borderMask := strings.Join([]string{
+		fmt.Sprintf(`<defs><mask id="%s" maskUnits="userSpaceOnUse" x="%d" y="%d" width="%d" height="%d">`,
+			maskID, targetShape.Pos.X, targetShape.Pos.Y-d2target.THREE_DEE_OFFSET, targetShape.Width+d2target.THREE_DEE_OFFSET, targetShape.Height+d2target.THREE_DEE_OFFSET,
+		),
+		fmt.Sprintf(`<rect x="%d" y="%d" width="%d" height="%d" fill="white"></rect>`,
+			targetShape.Pos.X, targetShape.Pos.Y-d2target.THREE_DEE_OFFSET, targetShape.Width+d2target.THREE_DEE_OFFSET, targetShape.Height+d2target.THREE_DEE_OFFSET,
+		),
+		fmt.Sprintf(`<path d="%s" style="%s;stroke:#000;fill:none;opacity:1;"/></mask></defs>`,
+			strings.Join(borderSegments, ""), borderStyle),
+	}, "\n")
+	// render the main hexagon without stroke and the border mask
+	mainShape := d2themes.NewThemableElement("polygon")
+	mainShape.X = float64(targetShape.Pos.X)
+	mainShape.Y = float64(targetShape.Pos.Y)
+	mainShape.Points = mainPointsPoly
+	mainShape.SetMaskUrl(maskID)
+	mainShapeFill, _ := d2themes.ShapeTheme(targetShape)
+	mainShape.Fill = mainShapeFill
+	mainShape.Stroke = color.None
+	mainShape.Style = targetShape.CSSStyle()
+	mainShapeRendered := mainShape.Render()
+
+	// render the side shapes in the darkened color without stroke and the border mask
+	var sidePoints []string
+	for _, v := range []d2target.Point{
+		{X: scale(targetShape.Width, 0.25) + d2target.THREE_DEE_OFFSET, Y: -Y_OFFSET},
+		{X: scale(targetShape.Width, 0.75) + d2target.THREE_DEE_OFFSET, Y: -Y_OFFSET},
+		{X: targetShape.Width + d2target.THREE_DEE_OFFSET, Y: scale(targetShape.Height, halfYFactor) - Y_OFFSET},
+		{X: scale(targetShape.Width, 0.75) + d2target.THREE_DEE_OFFSET, Y: targetShape.Height - Y_OFFSET},
+		{X: scale(targetShape.Width, 0.75), Y: targetShape.Height},
+		{X: targetShape.Width, Y: scale(targetShape.Height, halfYFactor)},
+		{X: scale(targetShape.Width, 0.75), Y: 0},
+		{X: scale(targetShape.Width, 0.25), Y: 0},
+	} {
+		sidePoints = append(sidePoints,
+			fmt.Sprintf("%d,%d", v.X+targetShape.Pos.X, v.Y+targetShape.Pos.Y),
+		)
+	}
+	// TODO make darker color part of the theme? or just keep this bypass
```

Comment:
bit of an ongoing debate hah, i think this is fine for now. https://github.com/terrastruct/d2/issues/896

Reply:
got it.